src/UserManager.php

```diff
@@ -0,0 +1,353 @@
+    $this->moduleHandler = $moduleHandler;
+    $this->logger = $logger;
+  }
+
+  /**
+   * Set options.
+   *
+   * @param array $options
+   *   The options.
+   */
+  public function setOptions(array $options) {
+    $this->options = $options;
+  }
+
+  /**
+   * Get user ids.
+   */
+  public function loadUserIds() {
+    $userIds = &drupal_static(__FUNCTION__);
+    if (!isset($userIds)) {
+      $userIds = $this->userStorage->getQuery()->execute();
+
+      // Handle modules.
+      $modules = $this->moduleConfig->get('modules');
+      if (is_array($modules)) {
+        $modules = array_flip(array_filter($this->moduleConfig->get('modules')));
+        foreach ($modules as $module) {
+          $moduleUsersIds = $this->getModuleUsersIds($module);
+          if (is_array($moduleUsersIds)) {
+            $userIds = array_intersect($userIds, $moduleUsersIds);
+          }
+        }
+      }
+
+      // Handle exclusions.
+      $exclusions = $this->moduleConfig->get('exclusions');
+      // Handle excluded roles.
+      $excludedRoles = $exclusions['roles'] ?? NULL;
+      if (is_array($excludedRoles) && !empty($excludedRoles)) {
+        $query = $this->userStorage->getQuery();
+        $group = $query->orConditionGroup();
+        foreach ($excludedRoles as $role) {
+          $group->condition('roles', $role);
+        }
+        $roleUserIds = $query
+          ->condition($group)
+          ->execute();
+        foreach ($roleUserIds as $userId) {
+          unset($userIds[$userId]);
+        }
+      }
+
+      // Handle excluded users.
+      $excludedUsers = $exclusions['users'] ?? NULL;
+      if (is_array($excludedUsers)) {
+        foreach ($excludedUsers as $userId) {
+          unset($userIds[$userId]);
+        }
+      }
+    }
+
+    return $userIds;
+  }
+
+  /**
+   * Mark users for deletion.
+   */
+  public function markUsersForDeletion() {
+    $this->validateConfig();
+
+    $userIds = $this->loadUserIds();
+    $this->setUserIds($userIds);
+    $this->logger->info($this->formatPlural(
+      count($userIds),
+      '1 user marked for deletion.',
+      '@count users marked for deletion.'
+    ));
+  }
+
+  /**
+   * Retain users.
+   *
+   * @param array $users
+   *   The users to retain.
+   */
+  public function retainUsers(array $users) {
+    $this->validateConfig();
+    $userIdClaim = $this->moduleConfig->get('api.user_id_claim');
+    $userIdField = $this->moduleConfig->get('general.user_id_field');
+
+    $this->logger->info($this->formatPlural(
+      count($users),
+      'Retaining one user.',
+      'Retaining @count users.'
+    ));
+    $userIds = $this->getUserIds();
+    if (is_array($userIds)) {
+      foreach ($users as $user) {
+        $userId = $user[$userIdClaim] ?? NULL;
+        if (NULL === $userId) {
+          throw new \RuntimeException(sprintf('Cannot get user id (%s)', $userIdClaim));
+        }
+        try {
+          $user = $this->loadUserByProperties([$userIdField => $userId]);
```

Comment:
Would be better to bulk load users, and not do individual queries. Please see: https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Entity%21EntityStorageInterface.php/function/EntityStorageInterface%3A%3AloadByProperties/8.2.x